ksqldb-execution/src/main/java/io/confluent/ksql/execution/function/udaf/KudafAggregator.java

```diff
@@ -52,9 +52,11 @@ public KudafAggregator(
 
   @Override
   public GenericRow apply(final K k, final GenericRow rowValue, final GenericRow aggRowValue) {
+    final GenericRow result = GenericRow.fromList(aggRowValue.values());
```

Comment:
Kafka Streams does not expect the aggregator to mutate its parameters. The streams code is passing in the "old value", which ksqlDB was then mutating and returning as the "new value". This meant, when then function returned, the old and new values matched.  This is obviously bad!

Code now takes a copy and mutates that.  There is a perf hit, obviously, but it's unavoidable.

Reply:
Not sure I understand -- why did the old code work, in that case? Or did something change on the Streams side recently?

Reply:
The old code works because we were never enabling the sending of old values. We now do, to avoid the spurious tombstones.

Reply:
Sorry, still not understanding. What was being sent before, if not the old values? Was this method even being called, previously?

Reply:
The processing nodes in the streams topology can optionally include the old/previous value, as well as the new/current value, to child nodes.  This is not on by default.  An upstream change to how table filters is handled means this is now turned on.

The streams code for aggregation looks something like:

```
V process(K key, Change<V> change) {
   // Get old value from store:
   final V oldAgg = store.get(key);

   // Undo any previous value:
  final T intermediateAgg = value.oldValue != null && oldAgg != null
    ? remove.apply(key, value.oldValue, oldAgg)
    : oldAgg;

   // Then add the new value
   final T newAgg;
   if (value.newValue != null) {
       final T initializedAgg = intermediateAgg == null
         ?  initializer.apply();
          : intermediateAgg;

       newAgg = add.apply(key, value.newValue, initializedAgg);
   } else {
       newAgg = intermediateAgg;
   }

   // update the store with the new value. & forard
   store.put(key, newAgg);
   tupleForwarder.maybeForward(key, newAgg, sendOldValues ? oldAgg : null);
}
```

The two calls: `remove.apply(key, value.oldValue, oldAgg)` and `add.apply(key, value.newValue, initializedAgg)` are calling out to ksqlDB code.  If these calls directly mutate the `oldAgg` or `initializedAgg` parameters passed, rather than creating copies, then the old and new values forwarded to child nodes will match. i.e. in `   tupleForwarder.maybeForward(key, newAgg, sendOldValues ? oldAgg : null)`, the parameters `newAgg` and `oldAgg` will have the same updated value, rather than `oldAgg` holding the previous value.  This breaks downstream processes, which expect the old and new value.

Previously the nodes weren't configured to send old values, so where just sending `null` for the old value and downstream could handle this correctly.